### PR TITLE
Add 'continue with user' option when credentials are already present

### PR DESCRIPTION
**Pull request by kirkbyo**
Closes https://github.com/andymatuschak/orbit/issues/179

<p align="center">

</p>

**What**
When the credentials are already present on the `iframe`'s domain, a "Continue with <email>" option is shown to confirm with the user that they are already authenticated.

<details>
<summary>Demo of Before and After behaviour</summary>

| Before | After |
| --- | --- |
| old_flow | new_flow |

</details>

**How**
If the `userRecord` is received before the user has entered their authentication credentials, we'll "double check" that they want to use the account we found. I added a new "ContinueWithUser` component to handle this behaviour to avoid bloating `SigninForm` further with conditionals. 

**Notes**
I will admit this new flow feels a bit "wonky". It feels weird to ask the user to "continue" without presenting them another option. I attempted to add a button that allows the user to "Use a different account", but to do this flow properly would involve reworking how the `overrideEmailAddress` is used right now. If we did want this behaviour, I'd be happy to take that on in a separate PR!

### CODE REVIEW

**Comment by andymatuschak** [comment]
Nice work, @kirkbyo! I'll comment at a high level before digging into the code changes.

At first, I thought: oh, boy, I guess we really have to make the sign in flow much more fleshed out to make this reasonable. To your point, it's weird not to have a second option!

But then I realized: hey, wait. In this flow, the user has _already entered_ their email address in the prior screen. This is quite different from other OAuth flows, where they really need to clarify that you _intend_ to sign in using the saved credential. In our case, if the saved credential's email address matches the one you entered, that's what you mean! This lets us simplify the UI language, and also means we don't need an alternative option.

So if the user's signed in, I think the secondary text can say: "Welcome back, `EMAIL`. We signed you in using a saved password." and the button can say "Continue".

Now, the other thing this makes me realize is that if the stored credential _doesn't_ match the email address they entered, we should discard it, sign out, and show the normal login interface. After all, they asked to sign in with a different address! A fancier flow might say "hey, wait, did you want to sign in with this account instead?" (people often forget which email address they use for things), but I think we can skip that for now.

**Comment by andymatuschak** [comment]
Oh, and probably we should `focus()` the button on mount, so that people can just hit return / space / whatever to continue.

**Review by andymatuschak — CHANGES_REQUESTED** [review]
(Oops, should have made my previous comment as a review to make it show up in your inbox appropriately, sorry)

**Comment by kirkbyo** [comment]
Makes sense! I have updated the auth flow with the changes you mentioned:
<p align="center">
Screen Shot 2021-06-02 at 12 11 22 PM
</p>

<details>
<summary>Demo</summary>

### Flow for when the `userRecord` and the `overrideEmailAddress` match
new_flowV2

### Flow for when the `userRecord` and the `overrideEmailAddress` __do not__ match

different_email

</details>

File: packages/ui/src/components/ContinueWithUser.tsx
Comment (andymatuschak) [inline packages/ui/src/components/ContinueWithUser.tsx:11]: Would you mind making this field required and throwing in the parent if `email` ends up `null`? If that happens I'd like to make sure it sounds alarms in Sentry. They won't be able to proceed anyway!

File: packages/app/src/signIn/SignInScreen.tsx
Comment (andymatuschak) [inline packages/app/src/signIn/SignInScreen.tsx:108]: I think this effect can be merged into the the one on line 123, which might make this state machine a bit clearer. Boy, effect hooks do start to fall apart when you get a bunch of them together like this.

File: packages/app/src/signIn/SignInScreen.tsx
Comment (andymatuschak) [inline packages/app/src/signIn/SignInScreen.tsx:121]: Ditto this one: maybe best to make one big effect which amounts to "we got a new user state (or account approval); let's see what needs doing." Wish I could see a better way to simplify this. I suppose we could reify the state machine more explicitly, but I'm not sure it's worth it.

File: packages/app/src/signIn/SignInScreen.tsx
Comment (andymatuschak) [inline packages/app/src/signIn/SignInScreen.tsx:91]: Nit: this name implies a boolean but it's really a more complex state. Perhaps `automaticLoginState`?

**Review by andymatuschak — APPROVED** [review]
Thanks for this, Ozzie! This patch reminds me that we’re probably just creating pain for ourselves by separating `ui` into a separate module like this. Perhaps at some point we’ll want to just move it into `app`. Or maybe just simple components like `Button` should be in a separate `ui` module? Or maybe a separate `style` module just for static bits like color / type hierarchy? Not sure.

Anyway, I made a few notes you might consider, but feel free to ignore all. Merge when you like.

**Comment by kirkbyo** [comment]
Done! I will merge once the tests go green.

> This patch reminds me that we’re probably just creating pain for ourselves by separating ui into a separate module like this. Perhaps at some point we’ll want to just move it into app. Or maybe just simple components like Button should be in a separate ui module? Or maybe a separate style module just for static bits like color / type hierarchy? Not sure.

My two cents: I personally like the idea of separating the UI from the app since it forces you to not encapsulate API logic within the UI and allows for testable components like `SigninForm`. However, I do think the folder could be separated into design system components (i.e. `Button`) vs. views (`SigninForm` or `ReviewArea`).  I worry that moving the design system bits into a separate package would become tedious at this stage since we'd now have two storybook instances.

**Comment by andymatuschak** [comment]
Those are all good points! 👍

**Pull request merged**
